request-cert/k8s_certificates.go

```diff
@@ -74,38 +97,100 @@ func kubernetesSigner(csrName string, csr []byte, wantServerAuth bool) ([]byte,
 		return nil, errors.Wrapf(err, "CertificateSigningRequest.Create(%s) failed", req.Name)
 	}
 
-	fmt.Printf("Request sent. To approve, run 'kubectl certificate approve %s'\n", req.Name)
+	fmt.Printf("Request sent, waiting for approval. To approve, run 'kubectl certificate approve %s'\n", req.Name)
+
+	// Build the watch request.
+	timeout := int64(watchTimeout.Seconds())
+	watchReq := types.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &timeout,
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", csrName).String(),
+	}
+
+	resultCh, err := client.Certificates().CertificateSigningRequests().Watch(watchReq)
+	if err != nil {
+		return nil, errors.Wrapf(err, "CertificateSigningRequest.Watch(%s) failed: %v", csrName)
+	}
 
-	// TODO(mberhault): we may want a timeout here, perhaps also retries.
-	lastLog := time.Now()
+	watchCh := resultCh.ResultChan()
+	// Loop until we have a cert, CSR is denied, or we timed out. Bail out on errors.
 	for {
-		resp, err = client.Certificates().CertificateSigningRequests().Get(req.Name, types.GetOptions{})
-		if err != nil {
-			return nil, errors.Wrapf(err, "CertificateSigningRequest.Get(%s) failed", req.Name)
+		select {
+		case event, ok := <-watchCh:
+			if !ok {
+				break
+			}
+
+			if event.Object.(*certificates.CertificateSigningRequest).UID != resp.UID {
+				// Wrong object.
```

Comment:
I'm not sure this can actually happen unless someone deleted the CSR and created one with the same name.